### ISSUE TITLE
illegal declarations should give error at declaration time, not later at instantiation time

### ISSUE THREAD

**Issue by arnetheduck** [nim-lang/Nim#3011]
In the following snippet, context on line 3 has a bad type, but instead the compiler complains that line 9 is wrong: `test.nim(9, 18) Error: invalid type: 'object' in this context: 'proc (): Value'` - this is obviously very confusing! if Value is not used in the function declaration, no error at all is given!

```nim
type
  Type = ref object 
    context: ref object

type
  Value = ref object
    typ: Type

proc x(): Value =
  nil
```

**Comment by Araq** [comment]
Yeah we know, duplicate of some other bug.

**Comment by BigBoneDaddy** [comment]
Maybe dupe of #2315?

**Comment by Araq** [comment]
BigBoneDaddy: Nope.

**Comment by rayman22201** [comment]
> Bug #3011 maybe? I tried to fix #3011 but it's nasty... It's not clear what is allowed and what not until it's used as a parameter/variable/etc.

@Araq, @timotheecour,

As Araq points out, it seems the way the compiler is designed makes this difficult to catch at declaration time.

As a compromise, maybe it's ok to have it detected at instantiation time, but create a better error message with more context information / better line numbers?
Similar to what I did in the PR: https://github.com/nim-lang/Nim/pull/10354

This is not perfect, because a bad type declaration that is never used will never be caught, but that seems much less harmful than a bad type declaration that is used but the programmer is unable to debug.

**Comment by ringabout** [comment]
Now it is even worse. It compiles since Nim v1.0

**Comment by bung87** [comment]
it compiles and the proc x return type is void.
Nim Compiler Version 1.7.1

**Comment by bung87** [comment]
!nim c
```nim
type
  Type = ref object 
    context: ref object

type
  Value = ref object
    typ: Type

proc x(): Value =
  nil
```
